mirgecom/flux.py

```diff
@@ -30,9 +33,94 @@
 THE SOFTWARE.
 """
 import numpy as np  # noqa
+from meshmode.dof_array import DOFArray
+from mirgecom.fluid import (
+    ConservedVars,
+    make_conserved
+)
 
 
-def lfr_flux(cv_tpair, f_tpair, normal, lam):
+def gradient_flux_central(u_tpair, normal):
+    r"""Compute a central flux for the gradient operator.
+
+    The central gradient flux, $\mathbf{h}$, of a scalar quantity $u$ is calculated
+    as:
+
+    .. math::
+
+        \mathbf{h}({u}^-, {u}^+; \mathbf{n}) = \frac{1}{2}
+        \left({u}^{+}+{u}^{-}\right)\mathbf{\hat{n}}
+
+    where ${u}^-, {u}^+$, are the scalar function values on the interior
+    and exterior of the face on which the central flux is to be calculated, and
+    $\mathbf{\hat{n}}$ is the *normal* vector.
+
+    *u_tpair* is the :class:`~grudge.trace_pair.TracePair` representing the scalar
+    quantities ${u}^-, {u}^+$. *u_tpair* may also represent a vector-quantity
+    :class:`~grudge.trace_pair.TracePair`, and in this case the central scalar flux
+    is computed on each component of the vector quantity as an independent scalar.
+
+    Parameters
+    ----------
+    u_tpair: :class:`~grudge.trace_pair.TracePair`
+        Trace pair for the face upon which flux calculation is to be performed
+    normal: numpy.ndarray
+        object array of :class:`~meshmode.dof_array.DOFArray` with outward-pointing
+        normals
+
+    Returns
+    -------
+    numpy.ndarray
+        object array of :class:`~meshmode.dof_array.DOFArray` with the flux for each
+        scalar component.
+    """
+    tp_avg = u_tpair.avg
+    tp_join = tp_avg
+    if isinstance(tp_avg, DOFArray):
+        return tp_avg*normal
+    elif isinstance(tp_avg, ConservedVars):
+        tp_join = tp_avg.join()
+
+    result = np.outer(tp_join, normal)
+
+    if isinstance(tp_avg, ConservedVars):
+        return make_conserved(tp_avg.dim, q=result)
+    else:
+        return result
+
+
+def divergence_flux_central(trace_pair, normal):
+    r"""Compute a central flux for the divergence operator.
+
+    The central divergence flux, $h$, is calculated as:
+
+    .. math::
+
+        h(\mathbf{v}^-, \mathbf{v}^+; \mathbf{n}) = \frac{1}{2}
+        \left(\mathbf{v}^{+}+\mathbf{v}^{-}\right) \cdot \hat{n}
+
+    where $\mathbf{v}^-, \mathbf{v}^+$, are the vectors on the interior and exterior
+    of the face across which the central flux is to be calculated, and $\hat{n}$ is
+    the unit normal to the face.
+
+    Parameters
+    ----------
+    trace_pair: :class:`~grudge.trace_pair.TracePair`
+        Trace pair for the face upon which flux calculation is to be performed
+    normal: numpy.ndarray
+        object array of :class:`~meshmode.dof_array.DOFArray` with outward-pointing
+        normals
+
+    Returns
+    -------
+    numpy.ndarray
+        object array of :class:`~meshmode.dof_array.DOFArray` with the flux for each
+        scalar component.
+    """
+    return trace_pair.avg@normal
```

Comment:
I wonder whether this routine is worth having (vs inlining the implementation), given the balance between code and documentation.

Reply:
IMO, it is important to explicitly delineate this as function with supporting documentation so that: a) the user/developer can easily read what it is and does, and b) so that users can *eventually* swap out for alternate implementations.

Reply:
I'm inclined to leave it for now. I think the implementation will grow somewhat with the `flux_func` parametrization stuff in #502. It would also maybe be nice to add an optional projection step as in, e.g., `inviscid_facial_flux`.